driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/AggregatePublisherImpl.java

```diff
@@ -16,87 +16,195 @@
+        this.allowDiskUse = allowDiskUse;
         return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> maxTime(final long maxTime, final TimeUnit timeUnit) {
-        wrapped.maxTime(maxTime, timeUnit);
+    public AggregatePublisher<T> batchSize(final int batchSize) {
+        super.batchSize(batchSize);
         return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> maxAwaitTime(final long maxAwaitTime, final TimeUnit timeUnit) {
-        wrapped.maxAwaitTime(maxAwaitTime, timeUnit);
+    public AggregatePublisher<T> maxTime(final long maxTime, final TimeUnit timeUnit) {
+        notNull("timeUnit", timeUnit);
+        this.maxTimeMS = TimeUnit.MILLISECONDS.convert(maxTime, timeUnit);
         return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> bypassDocumentValidation(final Boolean bypassDocumentValidation) {
-        wrapped.bypassDocumentValidation(bypassDocumentValidation);
+    public AggregatePublisher<T> maxAwaitTime(final long maxAwaitTime, final TimeUnit timeUnit) {
+        notNull("timeUnit", timeUnit);
+        this.maxAwaitTimeMS = TimeUnit.MILLISECONDS.convert(maxAwaitTime, timeUnit);
         return this;
     }
 
     @Override
-    public Publisher<Void> toCollection() {
-        return Publishers.publish(wrapped::toCollection);
+    public AggregatePublisher<T> bypassDocumentValidation(@Nullable final Boolean bypassDocumentValidation) {
+        this.bypassDocumentValidation = bypassDocumentValidation;
+        return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> collation(final Collation collation) {
-        wrapped.collation(collation);
+    public AggregatePublisher<T> collation(@Nullable final Collation collation) {
+        this.collation = collation;
         return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> comment(final String comment) {
-        wrapped.comment(comment);
+    public AggregatePublisher<T> comment(@Nullable final String comment) {
+        this.comment = comment;
         return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> hint(final Bson hint) {
-        wrapped.hint(hint);
+    public AggregatePublisher<T> hint(@Nullable final Bson hint) {
+        this.hint = hint;
         return this;
     }
 
     @Override
-    public AggregatePublisher<TResult> batchSize(final int batchSize) {
-        wrapped.batchSize(batchSize);
-        return this;
+    public Publisher<Void> toCollection() {
+        BsonDocument lastPipelineStage = getLastPipelineStage();
+        if (lastPipelineStage == null || !lastPipelineStage.containsKey("$out") && !lastPipelineStage.containsKey("$merge")) {
+            throw new IllegalStateException("The last stage of the aggregation pipeline must be $out or $merge");
+        }
+
+        return Mono.create(sink -> getExecutor().execute(operations.aggregateToCollection(pipeline, maxTimeMS, allowDiskUse,
```

Comment:
I will move this to `PublisherCreator` to be inline with all other calls to executor.execute.